ambari-server/src/main/java/org/apache/ambari/server/agent/stomp/dto/MetadataCluster.java

```diff
@@ -67,44 +71,31 @@ public static MetadataCluster emptyMetadataCluster() {
     return clusterLevelParams;
   }
 
-  public void setClusterLevelParams(SortedMap<String, String> clusterLevelParams) {
-    this.clusterLevelParams = clusterLevelParams;
-  }
-
   public SortedMap<String, SortedMap<String, String>> getAgentConfigs() {
     return agentConfigs;
   }
 
-  public void setAgentConfigs(SortedMap<String, SortedMap<String, String>> agentConfigs) {
-    this.agentConfigs = agentConfigs;
-  }
-
   public boolean updateServiceLevelParams(SortedMap<String, MetadataServiceInfo> update) {
-    boolean changed = false;
-    for (String key : update.keySet()) {
-      if (!serviceLevelParams.containsKey(key) || !serviceLevelParams.get(key).equals(update.get(key))) {
-        changed = true;
-        break;
-      }
-    }
-    if (changed) {
-      serviceLevelParams.putAll(update);
-    }
-    return changed;
+    return updateMapIfNeeded(serviceLevelParams, update);
   }
 
   public boolean updateClusterLevelParams(SortedMap<String, String> update) {
-    boolean changed = false;
-    for (String key : update.keySet()) {
-      if (!clusterLevelParams.containsKey(key) || !StringUtils.equals(clusterLevelParams.get(key), update.get(key))) {
-        changed = true;
-        break;
+    return updateMapIfNeeded(clusterLevelParams, update);
+  }
+
+  private <T> boolean updateMapIfNeeded(Map<String, T> mapToBeUpdated, Map<String, T> updatedMap) {
+    try {
+      LOCK.lock();
+      final boolean changed = !Objects.equals(mapToBeUpdated, updatedMap);
+
+      if (changed) {
+        mapToBeUpdated.clear();
```

Comment:
I agree with @adoroszlai that this is going to break partial metadata updates.
So we should either always send full metadata which will require additional modifications.
Or change logic here.

Reply:
I'll fix it soon